app/services/jbpm_process_service.rb

```diff
@@ -33,12 +33,27 @@ def process_options
       options = {
         'request'         => request,
         'request_context' => request.request_context.as_json,
-        'groups'          => [group].as_json
+        'groups'          => enhance_groups([group].as_json)
       }
     end
     options
   end
 
+  def enhance_groups(groups_json)
+    random_access_keys = []
+    groups_json.each do |group_json|
+      group_json['users']&.each do |user_json|
+        access_key = SecureRandom.hex(16)
+        user_json['random_access_key'] = access_key
+        full_name = "#{user_json['first_name']} #{user_json['last_name']}"
+        random_access_keys << RandomAccessKey.new(:access_key => access_key, :approver_name => full_name)
```

Comment:
I want to suggest a slight change to this logic.  It seems odd that the `RandomAccessKey` class does not create the key, but gets it passed to it.  That class should encapsulate the creation of the key.

Then this method changes to create the `RandomAccessKey` instance first by only passing the `approver_name` and you set `ser_json['random_access_key']` based on the `access_key` value from that instance.

```ruby
def enhance_groups(groups_json)
  groups_json.each do |group_json|
    group_json['users']&.each do |user_json|
      request.random_access_keys << random_access_key = RandomAccessKey.new(:approver_name => full_name)

      user_json['random_access_key'] = random_access_keys.access_key
      full_name = "#{user_json['first_name']} #{user_json['last_name']}"
    end
  end

  groups_json
end
```

Reply:
Moved the creation of key inside the model class.
I prefer to set `request.random_access_keys` outside the `do` block to optimize sql statements.